README.md
# url-run-command

<center>

[![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/Zuoqiu-Yingyi/url-run-command?include_prereleases&style=flat-square)](https://github.com/Zuoqiu-Yingyi/url-run-command/releases/latest)
[![GitHub Release Date](https://img.shields.io/github/release-date/Zuoqiu-Yingyi/url-run-command?style=flat-square)](https://github.com/Zuoqiu-Yingyi/url-run-command/releases/latest)
[![GitHub License](https://img.shields.io/github/license/Zuoqiu-Yingyi/url-run-command?style=flat-square)](https://github.com/Zuoqiu-Yingyi/url-run-command/blob/main/LICENSE)
[![GitHub last commit](https://img.shields.io/github/last-commit/Zuoqiu-Yingyi/url-run-command?style=flat-square)](https://github.com/Zuoqiu-Yingyi/url-run-command/commits/main)
![GitHub repo size](https://img.shields.io/github/repo-size/Zuoqiu-Yingyi/url-run-command?style=flat-square)
![hits](https://hits.b3log.org/Zuoqiu-Yingyi/url-run-command.svg)
[![GitHub all releases](https://img.shields.io/github/downloads/Zuoqiu-Yingyi/url-run-command/total?style=flat-square)](https://github.com/Zuoqiu-Yingyi/url-run-command/releases)

</center>

使用 URL 协议运行命令(Run the command using URL protocol)

---

## 原理 | PRINCIPLE

- 将自定义 URL 协议注册为 Windows 系统协议, 并调用一个具有如下功能的程序  
  Register the custom URL protocol as a Windows system protocol and call a program that has the following features.
  1. 对 URL 载荷进行 URL 解码, 获得指令  
     Decode the URL payload to get commands.
  2. 调用系统命令行解析器 `CMD` 或 `PowerShell` 运行指令  
     Call the system command line parser `CMD` or `PowerShell` to run the commands.

**注意**: 为了避免可能出现的 [远程命令/代码执行(Remote Command/Code Execution, RCE)](https://en.wikipedia.org/w/index.php?title=Remote_code_execution&redirect=no) 攻击, 请尽量使用足够长的随机字符串作为自定义协议名称  
**Note**: To avoid possible [RCE(Remote Command/Code Execution)](https://en.wikipedia.org/w/index.php?title=Remote_code_execution&redirect=no) attacks, try to use random strings long enough as custom URL scheme name.

## 开始 | START

### 快速开始 | QUICK START

1. 从 [Releases](https://github.com/Zuoqiu-Yingyi/url-run-command/releases) 中下载发行包并解压到自定义安装目录  
   Download the release package from [Releases](https://github.com/Zuoqiu-Yingyi/url-run-command/releases) and unzip it to a custom installation directory.
2. 运行安装目录中的 `install.bat` 批处理文件  
   Run the `install.bat` batch file in your installation directory.
3. 安装目录中将生成 `insert.reg` 与 `delete.reg` 两个注册表文件  
   `insert.reg` and `delete.reg` registry files are generated in the installation directory.
   - `insert.reg`: 安装时将要写入注册表的注册表项  
     `insert.reg`: The registry key that will be written to the registry at installation time.
   - `delete.reg`: 卸载时将要从注册表中删除的注册表项  
     `delete.reg`: The registry key that will be removed from the registry when uninstalled.
4. 允许对注册表的更改  
   Allow to change the registry.
5. 安装完成  
   The installation is complete.
6. 在 *Windows 资源管理器* 的地址栏中输入 `runcmd://systeminfo; pause;` 或 `runcmd:systeminfo; pause;` 并回车, 检查 URL 协议 `runcmd` 能否运行命令 `systeminfo; pause;` (查看系统信息后暂停)  
   Enter `runcmd://systeminfo; pause;` or `runcmd:systeminfo; pause;` in the address bar of Windows Explorer and enter to check if the URL scheme `runcmd` can run command `systeminfo; pause;` (pause after viewing system information).
7. 点击下方链接, 检查浏览器链接能否使用 `runcmd` 协议运行命令 `systeminfo; pause;`  
   Click on links below to check if the browser link can run command `systeminfo; pause;`   using the `runcmd` protocol.
   - [runcmd://systeminfo; pause;](https://diamondyuan.github.io/302/?location=runcmd%3A%2F%2Fsysteminfo%3B+pause%3B)
   - [runcmd:systeminfo; pause;](https://diamondyuan.github.io/302/?location=runcmd%3Asysteminfo%3B+pause%3B)

## 预览 | PREVIEW

![PING](./images/preview.png)

### 高级选项 | ADVANCED OPTIONS

在安装目录中使用如下命令查看命令行参数  
Use the following command in the installation directory to view command line parameters.

```powershell
PS > .\url-run.exe help
Active code page: 65001
SYNOPSIS
        url-run.exe <url> (-c|-p) [--debug] [--parser <path>]
        url-run.exe init [-n <scheme>] (-c|-p) [--debug] [--parser <path>]
        url-run.exe help

OPTIONS
        <url>       统一资源定位符 | Uniform Resource Locator, URL
        -c, --cmd   使用 CMD 解析器 | use CMD parser
        -p, --powershell, --pwsh
                    使用 PowerShell 解析器 | use PowerShell parser

        --debug     输出调试信息 | output debug information
        --parser    自定义命令解析器 | custom command parser
        <path>      自定义命令解析器路径 | path of custom command parser
        init        初始化注册表文件 | initialize the registry file
        -n, --name  自定义 URL 协议名称 | custom URL scheme name
        <scheme>    协议自定义名称 | name of custom scheme
        -c, --cmd   使用 CMD 解析器 | use CMD parser
        -p, --powershell, --pwsh
                    使用 PowerShell 解析器 | use PowerShell parser

        --debug     输出调试信息 | output debug information
        --parser    自定义命令解析器 | custom command parser
        <path>      自定义命令解析器路径 | path of custom command parser
        help        查看参数指南 | view the parameter guide
```

### 开发者选项 | DEVELOPER OPTIONS

```powershell
git clone https://github.com/Zuoqiu-Yingyi/url-run-command.git

cd url-run-command

g++ \
   src\main.cpp \
   -o dist\url-run \
   -g \
   -static

cd dist

.\url-run.exe init \
   --name "<自定义 URL 协议名 | name for custom URL scheme>" \
   --powershell \
   --parser "<自定义命令解析器路径 | path for custom command parser>"

start .\insert.reg
```

## 卸载 | UNINSTALL

1. 运行安装目录中的 `uninstall.bat` 批处理文件  
   Run the `uninstall.bat` batch file in your installation directory.
2. 允许对注册表的更改  
   Allow to change the registry.
3. 卸载完成  
   The uninstall is complete.

## 依赖 | DEPENDENCE

### 源码 | SOURCE

- [muellan/clipp: easy to use, powerful &amp; expressive command line argument parsing for modern C++ / single header / usage &amp; doc generation](https://github.com/muellan/clipp#overview)*[MIT License](https://github.com/muellan/clipp/blob/master/LICENSE)*

## 更改日志 | CHANGE LOG

[CHANGE LOG](./CHANGELOG.md)


CHANGELOG.md
# 更改日志 | CHANGE LOG

## v0.1.3 / 2021-04-06

- [v0.1.1 <=> v0.1.3](https://github.com/Zuoqiu-Yingyi/url-run-command/compare/v0.1.1...v0.1.3)
- 移除命令行中无效的引号 / Remove invalid quotes in the command line.

## v0.1.2 / 2021-04-06

- [v0.1.1 <=> v0.1.2](https://github.com/Zuoqiu-Yingyi/url-run-command/compare/v0.1.1...v0.1.2)
- 移除命令行中无效的引号 / Remove invalid quotes in the command line.

## v0.1.1 / 2021-11-28

- [v0.1.0 <=> v0.1.1](https://github.com/Zuoqiu-Yingyi/url-run-command/compare/v0.1.0...v0.1.1)
- 更改链接模式为静态链接 / Change the link mode to static link.

## v0.1.0 / 2021-11-27

- 项目初始化 / Project initialization.


src/mode.h
/**
 * @file mode.h
 * @Zuoqiu-Yingyi (49649786+Zuoqiu-Yingyi@users.noreply.github.com)
 * @brief
 * @version 0.1.0
 * @date 2021-11-27
 *
 * @copyright Copyright (c) 2021
 * @ref https://blog.csdn.net/sunjikui1255326447/article/details/88802927
 * @ref https://jingyan.baidu.com/article/9faa72318d67a5473c28cbe6.html
 */

#include "format.h"
#include "urldecode.h"
#include <cstdlib>
#include <direct.h>
#include <fstream>
#include <iostream>
#include <string>
#include <windows.h>

#ifndef URL2CMD_MODE
#    define URL2CMD_MODE

namespace url2cmd
{
enum class mode
{
    run,
    init,
    help
};

void run(const std::string &url, const std::string &parser, const bool cmd, const bool debug)
{
    system("cls");         // 清空命令行

    /* 提取 url 载荷 */
    auto payload = url.substr(url.find(':') + 1);
    auto begin   = payload.find_first_not_of("/");
    auto end     = payload.find_last_not_of("/");
    if (begin == std::string::npos || end == std::string::npos) throw "ERROR: URL 载荷为空! / URL payload is EMPTY!";

    auto command = urlDecode(payload.substr(begin, end - begin + 1));

    char cwd[1 << 15];
    _getcwd(cwd, 1 << 15);  // 获得程序运行目录

    std::cout << (cmd ? "" : "PS ") << cwd << "> " << command << std::endl;  // 打印将运行的命令

    if (debug) system("pause");

    if (cmd)
        system(stringFormat("%s /C %s", parser.c_str(), UTF82GBK(command.c_str()).c_str()).c_str());
    else
        system(stringFormat("%s -c %s", parser.c_str(), UTF82GBK(command.c_str()).c_str()).c_str());

    if (debug) system("pause");
}

void init(const std::string &scheme, std::string parser, const bool cmd, const bool debug)
{
    auto scheme_c_str = scheme.c_str();

    /* 将路径中的 '\' 替换为 '\\' */
    std::string pgmptr = _pgmptr;
    for (auto pos = pgmptr.find("\\"); pos != -1; pos = pgmptr.find("\\")) pgmptr.replace(pos, 1, "/");
    for (auto pos = pgmptr.find("/"); pos != -1; pos = pgmptr.find("/")) pgmptr.replace(pos, 1, "\\\\");
    for (auto pos = parser.find("\\"); pos != -1; pos = parser.find("\\")) parser.replace(pos, 1, "/");
    for (auto pos = parser.find("/"); pos != -1; pos = parser.find("/")) parser.replace(pos, 1, "\\\\");

    std::ofstream f_insert_reg, f_delete_reg;
    f_insert_reg.open("insert.reg");
    f_delete_reg.open("delete.reg");

    f_insert_reg << "Windows Registry Editor Version 5.00" << std::endl
              << std::endl
              << stringFormat("[HKEY_CURRENT_USER\\Software\\Classes\\%s]", scheme_c_str) << std::endl
              << stringFormat("@=\"URL:%s\"", scheme_c_str) << std::endl
              << "\"URL Protocol\"=\"\"" << std::endl
              << std::endl
              << stringFormat("[HKEY_CURRENT_USER\\Software\\Classes\\%s\\shell]", scheme_c_str) << std::endl
              << std::endl
              << stringFormat("[HKEY_CURRENT_USER\\Software\\Classes\\%s\\shell\\open]", scheme_c_str) << std::endl
              << std::endl
              << stringFormat("[HKEY_CURRENT_USER\\Software\\Classes\\%s\\shell\\open\\command]", scheme_c_str) << std::endl
              << stringFormat("@=\"\\\"%s\\\" %%1 %s %s --parser \\\"%s\\\"\"",
                              pgmptr.c_str(),
                              (cmd ? "--cmd" : "--powershell"),
                              (debug ? "--debug" : ""),
                              parser.c_str())
              << std::endl;
    f_delete_reg << "Windows Registry Editor Version 5.00" << std::endl
                << std::endl
                << stringFormat("[-HKEY_CURRENT_USER\\Software\\Classes\\%s]", scheme_c_str) << std::endl;

    f_insert_reg.close();
    f_delete_reg.close();
}

}  // namespace url2cmd

#endif


src/urldecode.h
/**
 * @file urldecode.h
 * @Zuoqiu-Yingyi (49649786+Zuoqiu-Yingyi@users.noreply.github.com)
 * @brief
 * @version 0.1.0
 * @date 2021-11-27
 *
 * @copyright Copyright (c) 2021
 *
 * @ref https://www.cnblogs.com/shike8080/archive/2017/01/18/6297210.html
 */
#include <assert.h>
#include <string>
#include <windows.h>

#ifndef URL2CMD_URLDECODE
#define URL2CMD_URLDECODE

namespace url2cmd
{
unsigned char hex2uint8(unsigned char x)
{
    unsigned char y;
    if (x >= '0' && x <= '9')
        y = x - '0';
    else if (x >= 'A' && x <= 'Z')
        y = x - 'A' + 10;
    else if (x >= 'a' && x <= 'z')
        y = x - 'a' + 10;
    else
        assert(0);
    return y;
}

std::string urlDecode(const std::string &str)
{
    auto   strTemp = str;
    auto   length  = str.length();
    size_t l       = 0;
    for (size_t i = 0; i < length; i++) {
        if (str[i] == '%') {
            assert(i + 2 < length);
            unsigned char high = hex2uint8((unsigned char)str[++i]);
            unsigned char low  = hex2uint8((unsigned char)str[++i]);
            strTemp[l++]       = (high << 4) + low;
        }
        else
            strTemp[l++] = str[i];
    }
    return strTemp.substr(0, l);
}

std::string UTF82GBK(const char *utf8)
{
    int      len  = MultiByteToWideChar(CP_UTF8, 0, utf8, -1, NULL, 0);
    wchar_t *wstr = new wchar_t[len + 1];
    memset(wstr, 0, len + 1);
    MultiByteToWideChar(CP_UTF8, 0, utf8, -1, wstr, len);
    len       = WideCharToMultiByte(CP_ACP, 0, wstr, -1, NULL, 0, NULL, NULL);
    char *str = new char[len + 1];
    memset(str, 0, len + 1);
    WideCharToMultiByte(CP_ACP, 0, wstr, -1, str, len, NULL, NULL);
    if (wstr) delete[] wstr;
    return str;
}

}  // namespace url2cmd

#endif


src/main.cpp
/**
 * @file main.cpp
 * @Zuoqiu-Yingyi (49649786+Zuoqiu-Yingyi@users.noreply.github.com)
 * @brief
 * @version 0.1.0
 * @date 2021-11-27
 *
 * @copyright Copyright (c) 2021
 *
 */
#include "clipp.h"
#include "mode.h"
#include <cstdlib>
#include <iostream>
#include <string>

void showArg(const int argc, char *argv[])
{
    for (size_t i = 0; i < argc; i++)  // 打印所有传入参数
        std::cout << i << '\t' << argv[i] << std::endl;
}

int main(int argc, char *argv[])
{
    system("chcp 65001");  // 切换字符集

    std::string   url;
    std::string   parser;
    std::string   scheme;
    bool          cmd      = false;
    bool          debug    = false;
    url2cmd::mode selected = url2cmd::mode::run;

    auto required_c_p  = clipp::required("-c", "--cmd").set(cmd).doc("使用 CMD 解析器 | use CMD parser") | clipp::required("-p", "--powershell", "--pwsh").doc("使用 PowerShell 解析器 | use PowerShell parser");
    auto option_debug  = clipp::option("--debug").set(debug).doc("输出调试信息 | output debug information");
    auto option_parser = clipp::option("--parser").doc("自定义命令解析器 | custom command parser") & clipp::value("path", parser).doc("自定义命令解析器路径 | path of custom command parser");

    auto run = (clipp::value("url", url).doc("统一资源定位符 | Uniform Resource Locator, URL"),
                required_c_p,
                option_debug,
                option_parser);

    auto init = (clipp::command("init").set(selected, url2cmd::mode::init).doc("初始化注册表文件 | initialize the registry file"),
                 clipp::option("-n", "--name").doc("自定义 URL 协议名称 | custom URL scheme name") & clipp::value("scheme", scheme).doc("协议自定义名称 | name of custom scheme"),
                 required_c_p,
                 option_debug,
                 option_parser);

    auto help = (clipp::command("help").set(selected, url2cmd::mode::help).doc("查看参数指南 | view the parameter guide"));

    auto cli = ((run | init | help));

    if (clipp::parse(argc, argv, cli)) {  // 参数解析成功
        if (debug) showArg(argc, argv);

        if (parser.empty()) {
            if (cmd)
                parser = "cmd";
            else
                parser = "powershell";
        }

        switch (selected) {
            case url2cmd::mode::run:
                url2cmd::run(url, parser, cmd, debug);
                break;
            case url2cmd::mode::init:
                url2cmd::init(scheme, parser, cmd, debug);
                std::cout << "注册表文件初始化成功 | registry file initialization was successful" << std::endl;
                break;
            case url2cmd::mode::help:
                std::cout << clipp::make_man_page(cli, argv[0]);
                break;
        }
    }
    else {
        std::cout << "ERROR: 参数解析失败! | Argument parse failed!" << std::endl;
        showArg(argc, argv);
        std::cout << clipp::make_man_page(cli, argv[0]);
        system("pause");
        return 0;
    }

    return 0;
}
